src/servermap.cpp

```diff
@@ -693,47 +706,50 @@ void ServerMap::loadBlock(std::string *blob, v3s16 p3d, MapSector *sector, bool
 			throw SerializationError("Invalid block data in database");
 		}
 	}
-}
-
-MapBlock* ServerMap::loadBlock(v3s16 blockpos)
-{
-	ScopeProfiler sp(g_profiler, "ServerMap: load block", SPT_AVG, PRECISION_MICRO);
-	bool created_new = (getBlockNoCreateNoEx(blockpos) == NULL);
 
-	v2s16 p2d(blockpos.X, blockpos.Z);
+	assert(block);
 
-	std::string ret;
-	dbase->loadBlock(blockpos, &ret);
-	if (!ret.empty()) {
-		loadBlock(&ret, blockpos, createSector(p2d), false);
-	} else if (dbase_ro) {
-		dbase_ro->loadBlock(blockpos, &ret);
-		if (!ret.empty()) {
-			loadBlock(&ret, blockpos, createSector(p2d), false);
-		}
-	} else {
-		return NULL;
-	}
+	if (created_new) {
+		ReflowScan scanner(this, m_emerge->ndef);
+		scanner.scan(block, &m_transforming_liquid);
 
-	MapBlock *block = getBlockNoCreateNoEx(blockpos);
-	if (created_new && (block != NULL)) {
 		std::map<v3s16, MapBlock*> modified_blocks;
 		// Fix lighting if necessary
 		voxalgo::update_block_border_lighting(this, block, modified_blocks);
 		if (!modified_blocks.empty()) {
-			//Modified lighting, send event
 			MapEditEvent event;
 			event.type = MEET_OTHER;
 			event.setModifiedBlocks(modified_blocks);
 			dispatchEvent(event);
 		}
 	}
-	return block;
+
+	if (save_after_load)
+		saveBlock(block);
+
+	// We just loaded it, so it's up-to-date.
+	block->resetModified();
```

Comment:
Should this really happen if block was already in memory?

Reply:
after deserializing the in-memory block will match the db, so yes

Reply:
Oh right, it overwrites the block.

But isn't overwriting the existing block wrong? Example:
* [emerge thread] checks for block existence (with env lock), it's not there
* [emerge thread] loads block from db
* [server thread] loads vmanip at block => loads block
* [server thread] sets node in block
* [emerge thread] inserts its old block from db with loadBlock => set-node from server thread is lost

Reply:
should be fixed, I also got rid of the goto